{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from copy import deepcopy\n",
    "from collections import Counter\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from src.data import data_io, dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_last_followup_status(patient_data):\n",
    "    if patient_data.get(\"vital_status\") == \"Dead\":\n",
    "        cancer_status = patient_data.get(\"person_neoplasm_cancer_status\", \"NA\")\n",
    "        primary_therapy_outcome = patient_data.get(\n",
    "            \"primary_therapy_outcome_success\", \"NA\"\n",
    "        )\n",
    "        return \"Dead\", cancer_status, primary_therapy_outcome\n",
    "    followups = list(patient_data.get(\"follow_ups\", {}).values())\n",
    "    if not followups:\n",
    "        return (\n",
    "            patient_data.get(\"vital_status\", \"NA\"),\n",
    "            \"NA\",\n",
    "            patient_data.get(\"primary_therapy_outcome_success\", \"NA\"),\n",
    "        )\n",
    "    last_followup = None\n",
    "    last_followup_days = -1\n",
    "    for followup in followups:\n",
    "        if followup.get(\"vital_status\") == \"Dead\":\n",
    "            # If the patient is dead, this must be the most recent followup\n",
    "            last_followup = followup\n",
    "            break\n",
    "        days = int(followup.get(\"days_to_last_followup\", 1))\n",
    "        if days > last_followup_days:\n",
    "            last_followup = followup\n",
    "            last_followup_days = days\n",
    "    vital_status = last_followup.get(\"vital_status\", \"NA\")\n",
    "    cancer_status = last_followup.get(\"person_neoplasm_cancer_status\", \"NA\")\n",
    "    primary_therapy_outcome = last_followup.get(\n",
    "        \"primary_therapy_outcome_success\", \"NA\"\n",
    "    )\n",
    "    return vital_status, cancer_status, primary_therapy_outcome\n",
    "\n",
    "\n",
    "def get_drug_treatment_responses(patient_data):\n",
    "    responses = set()\n",
    "    for treatment in patient_data.get(\"drugs\", []):\n",
    "        response = treatment.get(\"measure_of_response\")\n",
    "        if response:\n",
    "            responses.add(response)\n",
    "    return responses\n",
    "\n",
    "\n",
    "def get_stage_and_grade(patient_data: dict) -> tuple[str, str]:\n",
    "    stage = patient_data.get(\"clinical_stage\")\n",
    "    if not stage:\n",
    "        stage = patient_data.get(\"pathologic_stage\")\n",
    "    grade = patient_data.get(\"neoplasm_histologic_grade\")\n",
    "    return stage, grade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3702"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA_DIR = Path(\"../data\")\n",
    "CLINICAL_DIR = DATA_DIR / \"interim/clinical\"\n",
    "CLINICAL_DIR.mkdir(parents=True, exist_ok=True)\n",
    "data_io.unzip_files(DATA_DIR / \"raw/TCGA_clinical_bcr_xml.zip\", CLINICAL_DIR)\n",
    "clin_tcga = dataset.parse_clinical_data(CLINICAL_DIR)\n",
    "TCGA_rnaseq_counts_metadata = pd.read_csv(\"../data/raw/TCGA_rnaseq_counts_metadata.csv.gz\", index_col=0)\n",
    "\n",
    "tcga_coldata = pd.DataFrame(\n",
    "    [\n",
    "        [\n",
    "            clin_tcga[barcode][\"disease_code\"],\n",
    "            *get_stage_and_grade(clin_tcga[barcode]),\n",
    "            \"NA\",\n",
    "        ]\n",
    "        for barcode in clin_tcga\n",
    "    ],\n",
    "    columns=[\"disease_code\", \"stage\", \"grade\", \"stage_grade\"],\n",
    "    index=list(clin_tcga),\n",
    ")\n",
    "tcga_coldata[\"stage\"] = tcga_coldata[\"stage\"].str.replace(\n",
    "    r\"[^IV]\", \"\", regex=True\n",
    ")\n",
    "tcga_coldata = tcga_coldata[\n",
    "    tcga_coldata[\"stage\"].isin((\"I\", \"II\", \"III\", \"IV\"))\n",
    "]\n",
    "tcga_coldata[\"grade\"] = tcga_coldata[\"grade\"].str.replace(\"High Grade\", \"G3\")\n",
    "tcga_coldata[\"grade\"] = tcga_coldata[\"grade\"].str.replace(\n",
    "    r\"[^GXx\\d]\", \"\", regex=True\n",
    ")\n",
    "tcga_coldata = tcga_coldata[\n",
    "    tcga_coldata[\"grade\"].isin((\"G1\", \"G2\", \"G3\", \"G4\"))\n",
    "]\n",
    "tcga_coldata[\"stage_grade\"] = tcga_coldata[\"stage\"].str.cat(\n",
    "    tcga_coldata[\"grade\"], sep=\".\"\n",
    ")\n",
    "tcga_coldata = tcga_coldata[tcga_coldata.index.isin(set(TCGA_rnaseq_counts_metadata[\"cases.submitter_id\"]))]\n",
    "clin_tcga = {k: v for k, v in clin_tcga.items() if k in tcga_coldata.index}\n",
    "len(tcga_coldata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract platinum (PT) chemotherapy response groups\n",
    "*E.g. Carboplatin, Cisplatin, and Oxaliplatin*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "clin_pt_treated = {\n",
    "    k: deepcopy(v)\n",
    "    for k, v in clin_tcga.items()\n",
    "    if \"platin\" in str(v.get(\"drugs\", [])).lower()\n",
    "}\n",
    "for k, v in clin_pt_treated.items():\n",
    "    v[\"drugs\"] = [d for d in v[\"drugs\"].values() if \"platin\" in str(d).lower()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "CESC\n",
      "Subgroup counts:\n",
      "             Unlabeled  Sensitive  Resistant  Total\n",
      "stage_grade                                        \n",
      "I.G2                51         17          4    106\n",
      "I.G3                46         12          7     99\n",
      "II.G2               16         10          3     35\n",
      "II.G3               12         11          2     26\n",
      "III.G2              13          7          2     28\n",
      "III.G3              10          7          2     22\n",
      "IV.G2                4          3          3     11\n",
      "Total              152         67         23    327\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "STAD\n",
      "Subgroup counts:\n",
      "             Unlabeled  Sensitive  Resistant  Total\n",
      "stage_grade                                        \n",
      "II.G2               33          4          3     69\n",
      "II.G3               62         11          4    128\n",
      "III.G2              40          4          3     83\n",
      "III.G3              99         11          8    206\n",
      "Total              234         30         18    486\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "OV\n",
      "Subgroup counts:\n",
      "             Unlabeled  Sensitive  Resistant  Total\n",
      "stage_grade                                        \n",
      "II.G3                5          6          5     15\n",
      "III.G2              13          3         15     41\n",
      "III.G3             113         48        131    357\n",
      "Total              131         57        151    413\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "UCEC\n",
      "Subgroup counts:\n",
      "             Unlabeled  Sensitive  Resistant  Total\n",
      "stage_grade                                        \n",
      "I.G2                81          5          2    164\n",
      "I.G3               127         39          3    257\n",
      "III.G3              47         33         10    104\n",
      "IV.G3               20          5          3     43\n",
      "Total              275         82         18    568\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "BLCA\n",
      "Subgroup counts:\n",
      "             Unlabeled  Sensitive  Resistant  Total\n",
      "stage_grade                                        \n",
      "II.G3               90         12          8    188\n",
      "III.G3             109         22          5    223\n",
      "IV.G3              104         17         10    218\n",
      "Total              303         51         23    629\n",
      "\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "clin_pt_resistant = {}\n",
    "clin_pt_sensitive = {}\n",
    "\n",
    "for k in list(clin_pt_treated):\n",
    "    patient_data = clin_pt_treated[k]\n",
    "    stage, grade = get_stage_and_grade(patient_data)\n",
    "    if not stage or not grade or \"GX\" in grade:\n",
    "        continue\n",
    "    responses = get_drug_treatment_responses(patient_data)\n",
    "    if responses:\n",
    "        # Infer PT resistance based on documented responses\n",
    "        any_complete = any(\"Complete\" in r for r in responses)\n",
    "        any_partial = any(\"Partial\" in r for r in responses)\n",
    "        any_stable = any(\"Stable\" in r for r in responses)\n",
    "        any_progressive = any(\"Progressive\" in r for r in responses)\n",
    "        if (any_complete or any_partial) and not (any_stable or any_progressive):\n",
    "            clin_pt_sensitive[k] = patient_data\n",
    "        elif not (any_complete or any_partial) and (any_stable or any_progressive):\n",
    "            # Additional requirement for stage IV: \"Progressive Disease\" response only\n",
    "            # (stable might actually be a better than expected response in stage IV)\n",
    "            if \"IV\" in stage and any_stable:\n",
    "                continue\n",
    "            clin_pt_resistant[k] = patient_data\n",
    "    else:\n",
    "        # Infer PT resistance based on outcomes after treatment. To reduce risk of\n",
    "        # introducing bias, stage IV patients are excluded here due to poor prognosis.\n",
    "        # if \"IV\" not in stage:\n",
    "        vital_status, cancer_status, primary_therapy_outcome = get_last_followup_status(\n",
    "            patient_data\n",
    "        )\n",
    "        if (\n",
    "            vital_status == \"Alive\"\n",
    "            and cancer_status == \"TUMOR FREE\"\n",
    "            and \"Disease\" not in primary_therapy_outcome\n",
    "        ):\n",
    "            clin_pt_sensitive[k] = patient_data\n",
    "        elif (\n",
    "            vital_status == \"Dead\"\n",
    "            and cancer_status == \"WITH TUMOR\"\n",
    "            and \"Remission\" not in primary_therapy_outcome\n",
    "            and \"IV\" not in stage\n",
    "        ):\n",
    "            clin_pt_resistant[k] = patient_data\n",
    "\n",
    "# Filter out underrepresented stage/grade subgroups within each disease code\n",
    "# Require each subgroup to have at least 2 instances in each response group\n",
    "disease_codes = set(\n",
    "    [v[\"disease_code\"] for v in (clin_pt_resistant | clin_pt_sensitive).values()]\n",
    ")\n",
    "for disease_code in disease_codes:\n",
    "    resistant_barcodes = [\n",
    "        k for k, v in clin_pt_resistant.items() if v[\"disease_code\"] == disease_code\n",
    "    ]\n",
    "    sensitive_barcodes = [\n",
    "        k for k, v in clin_pt_sensitive.items() if v[\"disease_code\"] == disease_code\n",
    "    ]\n",
    "    resistant_subgroups = tcga_coldata.loc[resistant_barcodes]\n",
    "    sensitive_subgroups = tcga_coldata.loc[sensitive_barcodes]\n",
    "    sensitive_subgroups_counts = sensitive_subgroups[\"stage_grade\"].value_counts()\n",
    "    resistant_subgroups_counts = resistant_subgroups[\"stage_grade\"].value_counts()\n",
    "    exclude_subgroups = []\n",
    "    for group in set(sensitive_subgroups[\"stage_grade\"]).union(\n",
    "        set(resistant_subgroups[\"stage_grade\"])\n",
    "    ):\n",
    "        if (\n",
    "            group not in sensitive_subgroups_counts\n",
    "            or group not in resistant_subgroups_counts\n",
    "        ):\n",
    "            exclude_subgroups.append(group)\n",
    "        elif (\n",
    "            sensitive_subgroups_counts[group] < 2\n",
    "            or resistant_subgroups_counts[group] < 2\n",
    "        ):\n",
    "            exclude_subgroups.append(group)\n",
    "    exclude_barcodes_sen = sensitive_subgroups[\n",
    "        sensitive_subgroups[\"stage_grade\"].isin(exclude_subgroups)\n",
    "    ].index\n",
    "    exclude_barcodes_res = resistant_subgroups[\n",
    "        resistant_subgroups[\"stage_grade\"].isin(exclude_subgroups)\n",
    "    ].index\n",
    "    clin_pt_sensitive = {\n",
    "        k: v for k, v in clin_pt_sensitive.items() if k not in exclude_barcodes_sen\n",
    "    }\n",
    "    clin_pt_resistant = {\n",
    "        k: v for k, v in clin_pt_resistant.items() if k not in exclude_barcodes_res\n",
    "    }\n",
    "\n",
    "# Count frequencies of disease codes in each cohort\n",
    "resistant_counter = Counter(v[\"disease_code\"] for v in clin_pt_resistant.values())\n",
    "sensitive_counter = Counter(v[\"disease_code\"] for v in clin_pt_sensitive.values())\n",
    "\n",
    "# Keep only disease codes with more than 15 instances in each response group\n",
    "keep_codes = {\n",
    "    c\n",
    "    for c in resistant_counter\n",
    "    if resistant_counter[c] > 15 and sensitive_counter[c] > 15\n",
    "}\n",
    "\n",
    "# Filter out patients with disease codes not in keep_codes\n",
    "clin_pt_resistant = {\n",
    "    k: v for k, v in clin_pt_resistant.items() if v[\"disease_code\"] in keep_codes\n",
    "}\n",
    "clin_pt_sensitive = {\n",
    "    k: v for k, v in clin_pt_sensitive.items() if v[\"disease_code\"] in keep_codes\n",
    "}\n",
    "\n",
    "resistant_counter = Counter(v[\"disease_code\"] for v in clin_pt_resistant.values())\n",
    "sensitive_counter = Counter(v[\"disease_code\"] for v in clin_pt_sensitive.values())\n",
    "\n",
    "# Update TCGA coldata\n",
    "pt_cohorts_disease_codes = set([v[\"disease_code\"] for v in clin_pt_sensitive.values()])\n",
    "tcga_coldata_filt = tcga_coldata[\n",
    "    tcga_coldata[\"disease_code\"].isin(pt_cohorts_disease_codes)\n",
    "].copy()\n",
    "tcga_coldata_filt[\"PT_resistant\"] = \"NA\"\n",
    "tcga_coldata_filt.loc[clin_pt_sensitive.keys(), \"PT_resistant\"] = \"No\"\n",
    "tcga_coldata_filt.loc[clin_pt_resistant.keys(), \"PT_resistant\"] = \"Yes\"\n",
    "\n",
    "print(\"-\" * 80)\n",
    "for disease_code in pt_cohorts_disease_codes:\n",
    "    print(disease_code)\n",
    "    # sen_barcodes = [k for k, v in clin_pt_sensitive.items() if v[\"disease_code\"] == disease_code]\n",
    "    # res_barcodes = [k for k, v in clin_pt_resistant.items() if v[\"disease_code\"] == disease_code]\n",
    "    coldata_disease = tcga_coldata_filt[\n",
    "        tcga_coldata_filt[\"disease_code\"] == disease_code\n",
    "    ]\n",
    "    coldata_unlabeled = coldata_disease[coldata_disease[\"PT_resistant\"] == \"NA\"]\n",
    "    coldata_sensitive = coldata_disease[coldata_disease[\"PT_resistant\"] == \"No\"]\n",
    "    coldata_resistant = coldata_disease[coldata_disease[\"PT_resistant\"] == \"Yes\"]\n",
    "    unlabeled_counts = coldata_unlabeled[\"stage_grade\"].value_counts().sort_index()\n",
    "    sensitive_counts = coldata_sensitive[\"stage_grade\"].value_counts().sort_index()\n",
    "    resistant_counts = coldata_resistant[\"stage_grade\"].value_counts().sort_index()\n",
    "    unlabeled_counts = unlabeled_counts.loc[sensitive_counts.index]\n",
    "    unlabeled_proportions = unlabeled_counts / len(coldata_disease)\n",
    "    sensitive_proportions = sensitive_counts / len(coldata_disease)\n",
    "    resistant_proportions = resistant_counts / len(coldata_disease)\n",
    "    subgroup_counts_table = pd.DataFrame(\n",
    "        {\n",
    "            \"Unlabeled\": unlabeled_counts,\n",
    "            \"Sensitive\": sensitive_counts,\n",
    "            \"Resistant\": resistant_counts,\n",
    "            \"Total\": unlabeled_counts + unlabeled_counts + resistant_counts,\n",
    "        }\n",
    "    )\n",
    "    subgroup_counts_table.loc[\"Total\"] = subgroup_counts_table.sum(0)\n",
    "\n",
    "    print(\"Subgroup counts:\")\n",
    "    print(subgroup_counts_table)\n",
    "    print(\"\\n\" + \"-\" * 80)\n",
    "\n",
    "#######################################################################################\n",
    "# Notice the trend in the output for higher stages and grades to be labeled resistant.\n",
    "# Stage and grade are covariates for our target variable (inferred PT resistance).\n",
    "# To deal with this, we will adjust downstream analysis as follows:\n",
    "# - Add covariates stage and grade to DE analysis design formula.\n",
    "# - For binary classification, include stage and grade as categorical features.\n",
    "# - For binary classification, undersample the minority class per stage/grade subgroup.\n",
    "# - Use propensity scoring to evaluate binary classification models.\n",
    "# - For binary classification feature attributions, train additional models\n",
    "#   on random samples containing the same distribution of stage and grade within each\n",
    "#   class (classes = {sensitive, resistant}) as the labeled dataset, and use the\n",
    "#   feature attributions of models trained on pseudo samples as the baseline\n",
    "#   attribution of each feature. Subtract baseline feature attributions from the\n",
    "#   feature attribution values calculated for models trained on the labeled dataset.\n",
    "#######################################################################################\n",
    "\n",
    "### Downstream analyses (not necessarily in this order):\n",
    "# - PCA, UMAP, ICA\n",
    "# - Differential expression (DE) analysis\n",
    "# - Gene set enrichment analysis (GSEA)\n",
    "# - Weighted gene correlation network analysis (WGCNA)\n",
    "# - Binary classification and feature attribution explorations\n",
    "#   - Iterative random forest (iRF) using ranger (https://github.com/imbs-hl/ranger) for RF, Gini w/ p-values\n",
    "#     - stage and grade included as categorical variables and always considered for splitting. See `--catvars` and `--alwayssplitvars`\n",
    "#   - Top-scoring pairs (TSP) + Support vector machine (SVM)\n",
    "#   - L1 and L2 regularized logistic regression (Elastic net)\n",
    "#   - Pre-defined feature subsets\n",
    "#     - Hallmark gene sets from MSigDB\n",
    "#     - Matrisome, core matrisome (MSigDB: NABA_MATRISOME and NABA_CORE_MATRISOME)\n",
    "#     - PT sensitivity gene expression markers (links below)\n",
    "#       - Publication: https://www.nature.com/articles/s41388-021-02055-2\n",
    "#       - Database: http://ptrc-ddr.cptac-data-view.org/#/\n",
    "#       - Additional filters:\n",
    "#         - \"Source of the supporting data\" must include human cancer tissue\n",
    "#         - Source of supporting data cannot include TCGA (data leakage).\n",
    "#           - Check is performed by searching through articles by their PMIDs annotated in the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Prepare coldata and raw counts for each disease code cohort\n",
    "count_df = pd.read_parquet(DATA_DIR / \"raw/TCGA_rnaseq_counts.parquet\")\n",
    "genes_df = pd.read_csv(DATA_DIR / \"raw/genes.csv.gz\", index_col=0)\n",
    "assert np.all(genes_df.index == count_df.index)\n",
    "genes_keep = genes_df[\"gene_type\"] == \"protein_coding\"\n",
    "genes_df = genes_df[genes_keep]\n",
    "count_df = count_df.loc[genes_df.index]\n",
    "count_df.index = count_df.index.str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
    "\n",
    "# Use symbols instead of gene id. For duplicate symbols, keep only one w/ max total expression\n",
    "# sym = genes_df.loc[count_df.index, \"gene_name\"]\n",
    "# sym = sym.loc[count_df.assign(sum=count_df.sum(1), group=sym).groupby(\"group\")[\"sum\"].idxmax()]\n",
    "# count_df = count_df.loc[sym.index]\n",
    "# count_df.index = sym\n",
    "\n",
    "# Create count matrices for each disease code cohort\n",
    "barcodes_df = TCGA_rnaseq_counts_metadata\n",
    "for disease_code in pt_cohorts_disease_codes:\n",
    "    barcodes = tcga_coldata_filt[\n",
    "        tcga_coldata_filt[\"disease_code\"] == disease_code\n",
    "    ].index\n",
    "    count_df_disease = pd.DataFrame(\n",
    "        np.zeros((len(count_df), len(barcodes)), dtype=np.float32),\n",
    "        index=count_df.index,\n",
    "        columns=barcodes,\n",
    "    )\n",
    "    count_df_disease.index.name = \"gene\"\n",
    "    for barcode in barcodes:\n",
    "        fnames = barcodes_df[barcodes_df[\"cases.submitter_id\"] == barcode].index\n",
    "        counts = count_df[fnames].values\n",
    "        # If more than one expression profile available, use the one with more counts\n",
    "        count_df_disease[barcode] = counts[:, np.argmax(counts.sum(0))]\n",
    "\n",
    "    # Filter out genes with low expression (more than 50% of samples zero)\n",
    "    keep = (count_df_disease > 0).sum(1) > count_df_disease.shape[1] / 2\n",
    "    count_df_disease = count_df_disease[keep]\n",
    "\n",
    "    counts_dir = DATA_DIR / \"interim/TCGA/counts\"\n",
    "    counts_dir.mkdir(parents=True, exist_ok=True)\n",
    "    coldata_dir = DATA_DIR / \"interim/TCGA/coldata\"\n",
    "    coldata_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "    count_df_disease.to_csv(counts_dir / f\"{disease_code}.csv.gz\", compression=\"gzip\")\n",
    "    coldata = tcga_coldata_filt.loc[barcodes].drop(\n",
    "        columns=[\"disease_code\", \"stage_grade\"]\n",
    "    )\n",
    "    coldata.to_csv(coldata_dir / f\"{disease_code}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
